### PR TITLE
Handle more BinaryExpression cases

### PR DESCRIPTION
**Pull request by trueadm**
Release notes: none

Follow up to https://github.com/facebook/prepack/pull/1464. This adds serializer tests and fixes `+` binary expressions in pure mode.

### CODE REVIEW

File: test/serializer/pure-functions/BinaryExpressions2.js [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+}
+
+function additional2() {
+  return 42 < obj2.foo.bar;
+}
+
+inspect = function() {
+  let ret1 = additional1();
+  let ret2 = additional2();
+  return ret1 + ret2;
```
Comment (gaearon) [inline test/serializer/pure-functions/BinaryExpressions2.js]: We're adding booleans here. See my suggestion above.

File: test/serializer/pure-functions/BinaryExpressions2.js [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+let obj2 = global.__abstract ? __abstract({}, '({foo: {bar: {valueOf() { return 42; }}}})') : {foo: {bar: {valueOf() { return 42; }}}};
+
+if (global.__makePartial) {
+  __makePartial(obj1);
+  __makePartial(obj2);
+}
+if (global.__makeSimple) {
+  __makeSimple(obj1);
+  __makeSimple(obj2);
+}
```
Comment (gaearon) [inline test/serializer/pure-functions/BinaryExpressions2.js]: As far as I can tell, them being partial or simple is irrelevant to the code branch we're trying to test.

How about this instead?

```js
// additional functions
// abstract effects

let obj1 = global.__abstract ? __abstract('object', '({foo: {valueOf() { return 42; }}})') : {foo: {valueOf() { return 42; }}};
let obj2 = global.__abstract ? __abstract('object', '({foo: {bar: {valueOf() { return 42; }}}})') : {foo: {bar: {valueOf() { return 42; }}}};

function additional1() {
  return 42 < obj1.foo;
}

function additional2() {
  return 42 > obj2.foo.bar;
}

inspect = function() {
  let ret1 = additional1();
  let ret2 = additional2();
  return JSON.stringify({ ret1, ret2 });
}
```

File: test/serializer/pure-functions/BinaryExpressions.js [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+let obj2 = global.__abstract ? __abstract({}, '({foo: {bar: {valueOf() { return 42; }}}})') : {foo: {bar: {valueOf() { return 42; }}}};
+
+if (global.__makePartial) {
+  __makePartial(obj1);
+  __makePartial(obj2);
+}
+if (global.__makeSimple) {
+  __makeSimple(obj1);
+  __makeSimple(obj2);
+}
```
Comment (gaearon) [inline test/serializer/pure-functions/BinaryExpressions.js]: Same comment as below, I think we can lose it if we change `__abstract({}` to `__abstract('object'`. Simple/partial doesn't matter here.

**Comment by gaearon** [comment]
Here is an example that's broken with this assumption.

```js
// additional functions
// abstract effects

let obj1 = global.__abstract ? __abstract('object', '({valueOf() { this.x = 10; return 42; }})') : {valueOf() { this.x = 10; return 42; }};

function additional1() {
  var y = Object.create(obj1);
  y + ''; // mutation, but Prepack drops it
  return y.x;
}

function additional2() {
}

inspect = function() {
  return additional1();
}
```

Here is another example:

```js
// additional functions
// abstract effects

function additional1() {
  var x = {
    valueOf:
      (global.__abstract
        ? __abstract('function', '(function() { this.foo++; return 10; })') 
        : function() { this.foo++; return 10; }
      ),
    foo: 0
  };
  return x + '' + x.foo; // Prepack thinks it's safe to turn into "[object Object]0"
}

function additional2() {
}

inspect = function() {
  return additional1();
}
```

That's *probably* fine. I think I agree we should say that primitive methods like `toString` / `valueOf` / `@@toPrimitive` / etc are expected to be "purer" than other abstract external code, and it should probably be safe to assume they're idempotent and don't mutate their inputs.

This is a new assumption to me (I haven't seen it mentioned before). If @sebmarkbage is on board with this in pure mode let's continue.

**Comment by sebmarkbage** [comment]
@gaearon I think your examples are indicative of other bugs that already existed but not really related to this PR. Created a follow up here: #1476

**Pull request closed without merging**